ethash_opencl.go

```diff
@@ -560,7 +560,7 @@ func (c *OpenCLMiner) Search(block pow.Block, stop <-chan struct{}, index int) (
 					ds := C.uint64_t(c.dagSize)
 					// We verify that the nonce is indeed a solution by
 					// executing the Ethash verification function (on the CPU).
-					ret := C.ethash_light_compute_internal(c.ethash.Light.current.ptr, ds, hashToH256(headerHash), cn)
+					ret := C.ethash_light_compute_internal(c.ethash.Light.getCache(block.NumberU64()).ptr, ds, hashToH256(headerHash), cn)
```

Comment:
This is dangerous because of the finalizer. To keep the C pointer live during the call, you need to keep a reference to it, e.g.

```go
cache := c.ethash.Light.getCache(...)
C.ethash_light_compute_internal(cache.ptr, ...)
_ = cache // keep live
```

This is very subtle though, so I'd prefer a solution that shares this code with ethash.go.